src/Streamly/Internal/Unicode/Utf8/Type.hs

```diff
@@ -0,0 +1,474 @@
+                            return $ Fold.Partial arr1
+                        Unicode.ReplaceInvalid ->
+                            if isSurrogate c
+                            then replaceInvalid arr
+                            else do
+                                arr1 <- arr `snoc3` c
+                                return $ Fold.Partial arr1
+                | otherwise -> do
+                    arr1 <- arr `snoc4` c
+                    return $ Fold.Partial arr1
+
+    {-# INLINE dropInvalid #-}
+    dropInvalid arr = return $ Fold.Partial arr
+
+    {-# INLINE errorOnInvalid #-}
+    errorOnInvalid _ =
+        error
+            $ show "Streamly.Internal.Data.Stream.StreamD.encodeUtf8:"
+            ++ "Encountered a surrogate"
+
+    {-# INLINE replaceInvalid #-}
+    replaceInvalid arr = Fold.Partial <$> snoc3_ arr 239 191 189
+
+    {-# INLINE snoc2 #-}
+    snoc2 :: MArray.Array Word8 -> Char -> m (MArray.Array Word8)
+    snoc2 arr c =
+        assert (n >= 0x80 && n <= 0x07ff)
+            $ do
+                arr1 <- arr `MArray.snoc` x1
+                arr1 `MArray.snoc` x2
+
+        where
+
+        n = ord c
+        x1 = fromIntegral $ (n `shiftR` 6) + 0xC0
+        x2 = fromIntegral $ (n .&. 0x3F) + 0x80
+
+    {-# INLINE snoc3_ #-}
+    snoc3_ ::
+           MArray.Array Word8
+        -> Word8
+        -> Word8
+        -> Word8
+        -> m (MArray.Array Word8)
+    snoc3_ arr x1 x2 x3 = do
+        arr1 <- arr `MArray.snoc` x1
+        arr2 <- arr1 `MArray.snoc` x2
+        arr2 `MArray.snoc` x3
+
+    {-# INLINE snoc3 #-}
+    snoc3 :: MArray.Array Word8 -> Char -> m (MArray.Array Word8)
+    snoc3 arr c = assert (n >= 0x80 && n <= 0x07ff) (snoc3_ arr x1 x2 x3)
+
+        where
+
+        n = ord c
+        x1 = fromIntegral $ (n `shiftR` 12) + 0xE0
+        x2 = fromIntegral $ ((n `shiftR` 6) .&. 0x3F) + 0x80
+        x3 = fromIntegral $ (n .&. 0x3F) + 0x80
+
+    {-# INLINE snoc4 #-}
+    snoc4 :: MArray.Array Word8 -> Char -> m (MArray.Array Word8)
+    snoc4 arr c =
+        assert (n >= 0x80 && n <= 0x07ff)
+            $ do
+                arr1 <- arr `MArray.snoc` x1
+                arr2 <- arr1 `MArray.snoc` x2
+                arr3 <- arr2 `MArray.snoc` x3
+                arr3 `MArray.snoc` x4
+
+        where
+
+        n = ord c
+        x1 = fromIntegral $ (n `shiftR` 18) + 0xF0
+        x2 = fromIntegral $ ((n `shiftR` 12) .&. 0x3F) + 0x80
+        x3 = fromIntegral $ ((n `shiftR` 6) .&. 0x3F) + 0x80
+        x4 = fromIntegral $ (n .&. 0x3F) + 0x80
+
+{-# INLINE write #-}
+write :: MonadIO m => Fold m Char Utf8
+write = writeGeneric Unicode.ReplaceInvalid
+
+{-# INLINE read #-}
+read :: Unfold m Utf8 Char
+read = undefined
+
+--------------------------------------------------------------------------------
+-- Basic functions
+--------------------------------------------------------------------------------
+
+singleton :: Char -> Utf8
+singleton x = pack [x]
```

Comment:
`pack` may be inefficient because it may not fuse. We can use `fromStream . Stream.fromPure`.